### PR TITLE
Connections

### PR DESCRIPTION
**Pull request by petermarathas**
1d&3d: CNN, Max Pooling

### CODE REVIEW

File: bindsnet/network/topology.py [hunk 1]
```diff
@@ -356,6 +356,411 @@ def reset_state_variables(self) -> None:
...
+
+        width = (
+            input_depth - self.kernel_size[0] + 2 * self.padding[0]
+        ) / self.stride[0] + 1
+        height = (
+            input_width - self.kernel_size[1] + 2 * self.padding[1]
+        ) / self.stride[1] + 1
+        depth = (
+            input_height - self.kernel_size[2] + 2 * self.padding[2]
+        ) / self.stride[2] + 1
```
Comment (djsaunde) [inline bindsnet/network/topology.py:434]: I still don't understand this piece. You should be able to compute output width from input width, kernel size, and padding; ditto for height, right?

File: bindsnet/network/topology.py [hunk 2]
```diff
@@ -192,7 +192,7 @@ def normalize(self) -> None:
         ``self.norm``.
         """
         if self.norm is not None:
-            w_abs_sum = self.w.abs().sum(0).unsqueeze(0)
+            w_abs_sum = self.w.abs().sum(1).unsqueeze(1)
```
Comment (djsaunde) [inline bindsnet/network/topology.py:195]: What does this change do?

File: bindsnet/network/topology.py [hunk 3]
```diff
@@ -356,6 +356,411 @@ def reset_state_variables(self) -> None:
...
+
+    def reset_state_variables(self) -> None:
+        # language=rst
+        """
+        Contains resetting logic for the connection.
+        """
+        super().reset_state_variables()
+
+
+class Conv1dConnection(AbstractConnection):
```
Comment (djsaunde) [inline bindsnet/network/topology.py:525]: Can you rearrange the file such that it goes (in order) `Conv1dConnection`, `Conv2dConnection`, then `Conv3dConnection`?

File: bindsnet/network/topology.py [hunk 4]
```diff
@@ -612,7 +1106,7 @@ def normalize(self) -> None:
         """
         if self.norm is not None:
             w = self.w.view(self.source.n, self.target.n)
-            w *= self.norm / self.w.sum(0).view(1, -1)
+            w *= self.norm / self.w.sum(1).view(-1, 1)
```
Comment (djsaunde) [inline bindsnet/network/topology.py:1109]: Same question here: what does this do?

File: test/network/test_connections.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import torch
+import setuptools
```
Comment (djsaunde) [inline test/network/test_connections.py:2]: How are you running these tests? Have you tried `python -m pytest test/`? That might obviate the need of import (though I'm not sure why it's needed at all).

File: test/network/test_connections.py [hunk 6]
```diff
@@ -54,6 +55,32 @@ def test_transfer(self):
...
+            shape_a = [1]
+            shape_b = [1]
+            shape_a.extend([28] * dim)
+            shape_b.extend([26] * dim)
+            l_a = LIFNodes(shape=shape_a)
+            l_b = LIFNodes(shape=shape_b)
+            connection = conn_type(l_a, l_b, *args, **kwargs)
+            connection.normalize()
+            for fltr in range(connection.w.size(0)):
+                assert abs(connection.w[fltr].sum() - connection.norm) < 0.000001
```
Comment (djsaunde) [inline test/network/test_connections.py:81]: `torch.isclose` should work for this: https://pytorch.org/docs/stable/generated/torch.isclose.html

**Review by djsaunde — COMMENTED** [review]
Looks good overall, just a few questions here and there.

File: test/network/test_connections.py (hunk 5 above)
Comment (petermarathas) [reply test/network/test_connections.py]: Yeah, that did it.

File: bindsnet/network/topology.py (hunk 2 above)
Comment (petermarathas) [reply bindsnet/network/topology.py:195]: Testing `normalize` for `Connection` was failing (refer to [how I wrote the test](https://github.com/petermarathas/bindsnet-1/pull/2/files#diff-1aaf8ae7a2e615af7cb26ff2a9638b6eR57) to verify this). With this change, the test passed.

To give an example of the difference:
```
w = torch.tensor([
  [1,2],
  [3,4]
])
```
then
`w.sum(0) => tensor([4, 6])`
`w.sum(1) => tensor([3,  7])`

We want to sum each row, not each column, as each row corresponds to a filter in `self.w`, or so I understand it. `unsqueeze` in both cases just adds the dimension back where it was previously removed, so it should match whatever is used in `sum`.

File: bindsnet/network/topology.py (hunk 4 above)
Comment (petermarathas) [reply bindsnet/network/topology.py:1108]: This is similar to the normalize function above, just with a slightly different style to get there. Essentially the change is just swapping the dimensions being operated on, and that caused the normalize test to pass for LocalConnection. Again, please verify that I wrote the normalize testing function correctly.

File: bindsnet/network/topology.py (hunk 2 above)
Comment (djsaunde) [reply bindsnet/network/topology.py:195]: Nice, good find!

File: bindsnet/network/topology.py (hunk 4 above)
Comment (djsaunde) [reply bindsnet/network/topology.py:1108]: Okay, makes sense to me.

**Comment by djsaunde** [comment]
Just make sure to run the black formatter on the codebase!

**Comment by petermarathas** [comment]
> Just make sure to run the black formatter on the codebase!

@djsaunde I've tried running `black <file>` as well as `black -l 79 <file>` for both modified files. Both attempts seem to have failed the check, any idea why that would be?
